File: lib/java_buildpack/framework/riverbed_appinternals_agent.rb
# frozen_string_literal: true

# Cloud Foundry Java Buildpack
# Copyright 2013-2019 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'fileutils'
require 'java_buildpack/component/versioned_dependency_component'
require 'java_buildpack/framework'
module JavaBuildpack
  module Framework

    # Encapsulates the functionality for running the Riverbed Appinternals Agent support.
    class RiverbedAppinternalsAgent < JavaBuildpack::Component::VersionedDependencyComponent

      # Creates an instance
      #
      # @param [Hash] context a collection of utilities used the component
      def initialize(context)
        super(context)
        @uri = download_url(credentials, @uri)
      end

      # (see JavaBuildpack::Component::BaseComponent#compile)
      def compile
        JavaBuildpack::Util::Cache::InternetAvailability.instance.available(
          true, 'Downloading from Riverbed AppInternals Service Broker'
        ) do
          download_zip(false, @droplet.sandbox, @component_name)
        end
        @droplet.copy_resources
      rescue StandardError => e
        raise "Riverbed AppInternals download failed: #{e}"
      end

      # (see JavaBuildpack::Component::BaseComponent#release)
      def release
        @droplet.environment_variables
                .add_environment_variable('AIX_INSTRUMENT_ALL', 1)
                .add_environment_variable('DSA_PORT', dsa_port(credentials))
                .add_environment_variable('RVBD_AGENT_FILES', 1)
                .add_environment_variable('RVBD_AGENT_PORT', agent_port(credentials))
                .add_environment_variable('RVBD_JBP_VERSION', @version)

        @droplet.java_opts.add_agentpath(agent_path)

        return unless rvbd_moniker(credentials)

        @droplet.java_opts.add_system_property('riverbed.moniker', rvbd_moniker(credentials))
      end

      protected

      # (see JavaBuildpack::Component::VersionedDependencyComponent#supports?)
      def supports?
        @application.services.one_service?(FILTER)
      end

      private

      PROFILERURL = 'profilerUrlLinux'

      FILTER = /appinternals/.freeze

      private_constant :FILTER

      def agent_path
        @droplet.sandbox + 'agent/lib' + lib_name
      end

      def credentials
        service['credentials'] unless service.nil?
      end

      def service
        @application.services.find_service(FILTER)
      end

      def agent_port(credentials)
        credentials['rvbd_agent_port'] || 7073
      end

      def architecture
        `uname -m`.strip
      end

      def dsa_port(credentials)
        credentials['rvbd_dsa_port'] || 2111
      end

      def lib_name
        %w[x86_64 i686].include?(architecture) ? 'librpilj64.so' : 'librpilj.so'
      end

      def rvbd_moniker(credentials)
        credentials['rvbd_moniker'] || @configuration['rvbd_moniker']
      end

      def download_url(credentials, default_url)
        (credentials[PROFILERURL] unless credentials.nil?) || default_url
      end

    end
  end
end

File: src/java/frameworks/riverbed_appinternals_agent.go
// Cloud Foundry Java Buildpack
// Copyright 2013-2021 the original author or authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package frameworks

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

// RiverbedAppInternalsAgentFramework represents the Riverbed AppInternals agent framework
type RiverbedAppInternalsAgentFramework struct {
	context   *Context
	agentPath string
}

// NewRiverbedAppInternalsAgentFramework creates a new Riverbed AppInternals agent framework instance
func NewRiverbedAppInternalsAgentFramework(ctx *Context) *RiverbedAppInternalsAgentFramework {
	return &RiverbedAppInternalsAgentFramework{context: ctx}
}

// Detect checks if Riverbed AppInternals agent should be enabled
func (r *RiverbedAppInternalsAgentFramework) Detect() (string, error) {
	// Check for riverbed-appinternals service binding
	if r.hasServiceBinding() {
		r.context.Log.Debug("Riverbed AppInternals agent framework detected via service binding")
		return "riverbed-appinternals-agent", nil
	}

	r.context.Log.Debug("Riverbed AppInternals agent: no service binding found")
	return "", nil
}

// Supply downloads and installs the Riverbed AppInternals agent
func (r *RiverbedAppInternalsAgentFramework) Supply() error {
	r.context.Log.BeginStep("Installing Riverbed AppInternals agent")

	// Get dependency from manifest
	dep, err := r.context.Manifest.DefaultVersion("riverbed-appinternals-agent")
	if err != nil {
		return fmt.Errorf("unable to find Riverbed AppInternals agent in manifest: %w", err)
	}

	// Install the agent
	agentDir := filepath.Join(r.context.Stager.DepDir(), "riverbed_appinternals_agent")
	if err := r.context.Installer.InstallDependency(dep, agentDir); err != nil {
		return fmt.Errorf("failed to install Riverbed AppInternals agent: %w", err)
	}

	// Find the installed agent directory (contains lib/rvbd-agent.jar)
	agentJarPath := filepath.Join(agentDir, "lib", "rvbd-agent.jar")
	if _, err := os.Stat(agentJarPath); err != nil {
		return fmt.Errorf("Riverbed AppInternals agent JAR not found after installation: %w", err)
	}
	r.agentPath = agentJarPath

	r.context.Log.Info("Riverbed AppInternals agent %s installed", dep.Version)
	return nil
}

// Finalize configures the Riverbed AppInternals agent
func (r *RiverbedAppInternalsAgentFramework) Finalize() error {
	if r.agentPath == "" {
		return nil
	}

	r.context.Log.BeginStep("Configuring Riverbed AppInternals agent")

	// Convert staging path to runtime path
	relPath, err := filepath.Rel(r.context.Stager.DepDir(), r.agentPath)
	if err != nil {
		return fmt.Errorf("failed to determine relative path for Riverbed AppInternals agent: %w", err)
	}
	runtimeJarPath := filepath.Join("$DEPS_DIR/0", relPath)

	// Get credentials from service binding
	credentials := r.getCredentials()

	// Build all JAVA_OPTS options
	var opts []string
	opts = append(opts, fmt.Sprintf("-javaagent:%s", runtimeJarPath))

	// Configure moniker (application name)
	moniker := credentials.Moniker
	if moniker == "" {
		moniker = r.getApplicationName()
	}
	if moniker != "" {
		opts = append(opts, fmt.Sprintf("-Drvbd.moniker=%s", moniker))
	}

	// Configure analysis server
	if credentials.AnalysisServer != "" {
		opts = append(opts, fmt.Sprintf("-Drvbd.analysis.server=%s", credentials.AnalysisServer))
	}

	// Write all options to .opts file
	javaOpts := strings.Join(opts, " ")
	if err := writeJavaOptsFile(r.context, 37, "riverbed_appinternals_agent", javaOpts); err != nil {
		return fmt.Errorf("failed to write JAVA_OPTS for Riverbed AppInternals: %w", err)
	}

	r.context.Log.Info("Riverbed AppInternals agent configured")
	return nil
}

// hasServiceBinding checks if there's a riverbed-appinternals service binding
func (r *RiverbedAppInternalsAgentFramework) hasServiceBinding() bool {
	vcapServices, err := GetVCAPServices()
	if err != nil {
		r.context.Log.Debug("Failed to parse VCAP_SERVICES: %s", err.Error())
		return false
	}

	// Check for Riverbed AppInternals service binding via multiple methods
	if vcapServices.HasService("riverbed-appinternals") ||
		vcapServices.HasService("appinternals") ||
		vcapServices.HasTag("riverbed") ||
		vcapServices.HasTag("appinternals") ||
		vcapServices.HasServiceByNamePattern("riverbed") ||
		vcapServices.HasServiceByNamePattern("appinternals") {
		return true
	}

	return false
}

// RiverbedCredentials holds Riverbed AppInternals credentials
type RiverbedCredentials struct {
	Moniker        string
	AnalysisServer string
}

// getCredentials retrieves Riverbed credentials from service binding
func (r *RiverbedAppInternalsAgentFramework) getCredentials() RiverbedCredentials {
	creds := RiverbedCredentials{}

	vcapServices, err := GetVCAPServices()
	if err != nil {
		return creds
	}

	// Try to find service by exact label first
	var service *VCAPService
	if svc := vcapServices.GetService("riverbed-appinternals"); svc != nil {
		service = svc
	} else if svc := vcapServices.GetService("appinternals"); svc != nil {
		service = svc
	} else {
		// Fall back to pattern matching for user-provided services
		service = vcapServices.GetServiceByNamePattern("riverbed")
		if service == nil {
			service = vcapServices.GetServiceByNamePattern("appinternals")
		}
	}

	if service == nil {
		return creds
	}

	// Extract moniker (application name) - try multiple key variations
	if moniker, ok := service.Credentials["moniker"].(string); ok {
		creds.Moniker = moniker
	} else if moniker, ok := service.Credentials["rvbd_moniker"].(string); ok {
		creds.Moniker = moniker
	}

	// Extract analysis server - try multiple key variations
	if server, ok := service.Credentials["analysis_server"].(string); ok {
		creds.AnalysisServer = server
	} else if server, ok := service.Credentials["analysisServer"].(string); ok {
		creds.AnalysisServer = server
	} else if server, ok := service.Credentials["rvbd_analysis_server"].(string); ok {
		creds.AnalysisServer = server
	}

	return creds
}

// getApplicationName returns the application name from VCAP_APPLICATION
func (r *RiverbedAppInternalsAgentFramework) getApplicationName() string {
	vcapApp := os.Getenv("VCAP_APPLICATION")
	if vcapApp == "" {
		return ""
	}

	var appData map[string]interface{}
	if err := json.Unmarshal([]byte(vcapApp), &appData); err != nil {
		return ""
	}

	if name, ok := appData["application_name"].(string); ok {
		return name
	}

	return ""
}

